src/js/api/core-bundle-loader.js

```diff
@@ -0,0 +1,135 @@
+import Item from 'playlist/item';
+import { fixSources } from 'playlist/playlist';
+import { SupportsMatrix } from 'providers/providers-supported';
+import registerProvider from 'providers/providers-register';
+import { ControlsLoader } from 'controller/controls-loader';
+import { SETUP_ERROR_LOADING_CORE_JS } from 'api/errors';
+import { loadCore, chunkLoadErrorHandler } from 'api/core-loader';
+
+let bundlePromise = null;
+
+export const bundleContainsProviders = {};
+
+export function loadCoreBundle(model) {
+    if (!bundlePromise) {
+        bundlePromise = selectBundle(model);
+    }
+    return bundlePromise;
+}
+
+function selectBundle(model) {
+    const controls = model.get('controls');
+    const polyfills = requiresPolyfills();
+    const html5Provider = requiresProvider(model, 'html5');
+
+    if (controls && polyfills && html5Provider) {
+        return loadControlsPolyfillHtml5Bundle();
+    }
+    if (controls && html5Provider) {
+        return loadControlsHtml5Bundle();
+    }
+    if (controls && polyfills) {
+        return loadControlsPolyfillBundle();
+    }
+    if (controls) {
+        return loadControlsBundle();
+    }
+    return loadWebCore();
+}
+
+function requiresPolyfills() {
+    const IntersectionObserverEntry = window.IntersectionObserverEntry;
+    return !IntersectionObserverEntry ||
+        !('IntersectionObserver' in window) ||
+        !('intersectionRatio' in IntersectionObserverEntry.prototype);
+}
+
+function requiresProvider(model, providerName) {
+    const playlist = model.get('playlist');
+    if (Array.isArray(playlist) && playlist.length) {
+        const sources = fixSources(Item(playlist[0]), model);
+        for (let i = 0; i < sources.length; i++) {
+            const source = sources[i];
+            const providersManager = model.getProviders();
+            for (let j = 0; j < SupportsMatrix.length; j++) {
+                const provider = SupportsMatrix[j];
+                if (providersManager.providerSupports(provider, source)) {
```

Comment:
if we had two providers that support the source and the first is not the passed in `providerName` this logic would return false despite the next loop potentially returning true. is this expected? little unclear on this logic

Reply:
The logic isn't changing here, it's just being reorganized to remove code-splitting output in the headless player. The headless player outputs jwplayer.core.js but none of the bundled alternatives here, because this module is rerouted to empty output in the headless webpack target.

To answer your question the only `providerName` we would pass in, would be one we have a bundle for. If the first provider supported supported is not a match for a bundle we might load, we will not load that bundle because the player will select the first provider and load it on demand at a later point. If we do have a match and load the bundle, the provider will not need to be loaded later.

Reply:
Got it, thanks for the explanation 👍